File: plottingTools/coffea_vs_RDF/makeHistogramsRDF.py

from collections import OrderedDict

import ROOT
ROOT.gROOT.SetBatch(True)

import time



LUMI = 21071.0+38654.0
XSection = 0.01891


# Used to define root files to read
samplesPath = "/eos/home-f/fleble/SVJ/data/production/102X/tchannel/mMed-3000_mDark-20_rinv-0.3_alpha-peak_yukawa-1_13TeV-madgraphMLM-pythia8/NANOAODJMAR/"
filesNumber = [ i for i in range(1,101) ]


# Define the binning of the different variables to histogram
ranges = {
#    "nGoodFatJet"          : (20 , 0 , 20  ),
    "GoodFatJet_pt"        : (200, 0 , 2000),
}

# Define the objects and variables to build up the variables to histogram
objDefinitions = OrderedDict()
o = objDefinitions
o["GoodFatJet"]       = "abs(FatJet_eta) < 2.4 && FatJet_pt > 200"
#o["nGoodFatJet"]      = "Sum(GoodFatJet)"
o["GoodFatJet_pt"]    = "FatJet_pt[GoodFatJet]"



# Book a histogram for a specific variable
def bookHistogram(df, variable, range_, weight):
    return df.Histo1D(ROOT.ROOT.RDF.TH1DModel(variable, variable, range_[0], range_[1], range_[2]), variable, weight)


# Write a histogram with a given name to the output ROOT file
def writeHistogram(h, name):
    h.SetName(name)
    h.Write()



def main():

    # Set up multi-threading capability of ROOT
    ROOT.ROOT.EnableImplicitMT(8)

    # Variables to histogram
    variables = ranges.keys()

    # Create output file
    tfile = ROOT.TFile("./histograms.root", "RECREATE")

    # Initialise event counter
    nGenEvts = 0

    # Dict to store histograms
    hists = {}


    # Loop through datasets and produce histograms of variables
    firstPass = True
    for fn in filesNumber:

        fileName = samplesPath + str(fn) + ".root"

        # Load dataset
        df = ROOT.ROOT.RDataFrame("Events", fileName)
        dfRun = ROOT.ROOT.RDataFrame("Runs", fileName)

        for obj, definition in objDefinitions.items():
            df = df.Define(obj, definition)

        # Increment the number of event processed
        nGenEvts += dfRun.Mean("genEventSumw_").GetValue()
        # nGenEvts += df.Sum("genWeight").GetValue()

        # Book histograms
        for variable in variables:
            weight = "genWeight"
            if firstPass:
                hists[variable] = bookHistogram(df, variable, ranges[variable], weight)
                firstPass = False
            else:
                h = bookHistogram(df, variable, ranges[variable], weight)
                hists[variable].Add(h.GetValue())


    # Normalise histograms
    for variable in variables:
        hists[variable].Scale( XSection*LUMI/nGenEvts )

    # Write histograms to output file
    for variable in variables:
        tfile.cd()
        writeHistogram(hists[variable], "{}".format(variable))

    tfile.Close()


if __name__ == "__main__":
    tstart = time.time()
    main()
    elapsed = time.time() - tstart
    print("Elapsed time: %d s" %elapsed)

File: plottingTools/coffea_vs_RDF/makeHistogramsRDF.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <future>

#include "ROOT/RDataFrame.hxx"

using namespace std;
using namespace ROOT;
using namespace ROOT::RDF;

const double LUMI = 21071.0+38654.0;
const double XSection = 0.01891;

// Used to define root files to read
const string fileName = "../../../data/t_channel/NANOAODJMAR/merged/1.root";

// Define the binning of the different variables to histogram
const map<string, tuple<int, double, double>> ranges = {
  //  "nGoodFatJet"          : {20 , 0 , 20  },
  {"GoodFatJet_pt", {200, 0 , 2000}},
};

// Define the objects and variables to build up the variables to histogram
const map<string, string> objDefinitions = {
  {"GoodFatJet",  "abs(FatJet_eta) < 2.4 && FatJet_pt > 200"},
  //o["nGoodFatJet"]      = "Sum(GoodFatJet)"
  {"GoodFatJet_pt",   "FatJet_pt[GoodFatJet]"},
};


template<class T> double duration(T t0,T t1)
{
  auto elapsed_secs = t1-t0;
  typedef chrono::duration<float> float_seconds;
  auto secs = chrono::duration_cast<float_seconds>(elapsed_secs);
  return secs.count();
}

inline chrono::time_point<std::chrono::steady_clock> now()
{
  return chrono::steady_clock::now();
}

// Book a histogram for a specific variable
RResultPtr<TH1D> bookHistogram(RInterface<ROOT::Detail::RDF::RLoopManager, void> df,
                               string variable, tuple<int, double, double> range_, string weight)
{
  auto histModel = TH1DModel(variable.c_str(), variable.c_str(), get<0>(range_), get<1>(range_), get<2>(range_));
  return df.Histo1D(histModel, variable, weight);
}


// Write a histogram with a given name to the output ROOT file
void writeHistogram(RResultPtr<TH1D> h, string name)
{
  h->SetName(name.c_str());
  h->Write();
}


//void makeHistogramsRDF()
int main()
{
  auto tstart = now();
  ROOT::EnableImplicitMT(8);
  
  // Load dataset
  auto df = RDataFrame("Events", fileName);
  
  string firstKey = objDefinitions.begin()->first;
  string firstValue = objDefinitions.begin()->second;
  
  auto dfAfterCuts = df.Define(firstKey, firstValue);
  bool first = true;
  
  for(auto &[obj, definition] : objDefinitions){
    if(first){
      first = false;
      continue;
    }
    dfAfterCuts = dfAfterCuts.Define(obj, definition);
  }
  
  // Create output file
  TFile *tfile = new TFile("./histograms_cpp.root", "RECREATE");
  tfile->cd();
  
  // Get mean of event weight sums (why?)
  auto dfRun = RDataFrame("Runs", fileName);
  int nGenEvts = dfRun.Mean("genEventSumw_").GetValue();
  
  // Book and fill histograms
  for(auto &[variable, range] : ranges){
    string weight = "genWeight";
    auto hist = bookHistogram(dfAfterCuts, variable, range, weight);
    hist->Scale( XSection*LUMI/nGenEvts );
    writeHistogram(hist, variable);
  }
  
  tfile->Close();
  
  auto elapsed = duration(tstart, now());
  cout<<"Elapsed time: "<<elapsed<<endl;
}


